conda/core/solve.py

```diff
@@ -1029,6 +1039,250 @@ def _prepare(self, prepared_specs):
         return self._index, self._r
 
 
+class LibSolvSolver(Solver):
+    """
+    This alternative Solver logic wraps ``libsolv`` through their Python bindings.
+
+    We mainly replace ``Solver.solve_final_state``, quite high in the abstraction
+    tree, which means that ``conda.resolve`` and ``conda.common.logic`` are no longer
+    used here. All those bits are replaced by the logic implemented in ``libsolv`` itself.
+
+    This means that ``libsolv`` governs processes like:
+
+    - Building the SAT clauses
+    - Pruning the repodata (?) - JRG: Not sure if this happens internally or at all.
+    - Prioritizing different aspects of the solver (version, build strings, track_features...)
+    """
+
+    def __init__(self, prefix, channels, subdirs=(), specs_to_add=(), specs_to_remove=(),
+                 repodata_fn=REPODATA_FN, command=NULL):
+        super().__init__(prefix, channels, subdirs=subdirs, specs_to_add=specs_to_add,
+                         specs_to_remove=specs_to_remove, repodata_fn=repodata_fn,
+                         command=command)
+        self._state = {}
+
+    def solve_for_transaction(self, update_modifier=NULL, deps_modifier=NULL, prune=NULL,
+                              ignore_pinned=NULL, force_remove=NULL, force_reinstall=NULL,
+                              should_retry_solve=False):
+        # Logic heavily based on Mamba's implementation:
+        # https://github.com/mamba-org/mamba/blob/master/mamba/mamba.py#L289 (solver parts)
+
+        print("------ USING EXPERIMENTAL MAMBA INTEGRATIONS ------")
+
+        try:
+            # 1. Populate repos with installed packages
+            self._setup_state()
+            # 2. Create solver and needed flags, tasks and jobs
+            self._configure_solver()
+            # 3. Run the SAT solver
+            self._run_solver()
+            # 5. Export back to conda
+            return self._build_transaction()
+        finally:
+            # 6. Clean
+            self._state.clear()
```

Comment:
Using `self._state` as an internal ephemeral argument is kind of anti-pattern. Wouldn't this be more idiomatic by using the state returned by `_setup_state` and passing it to the other functions?

Reply:
I was mimicking their `SolverStateContainer` class behaviour (not a fan either), but without knowing which things we will have to move around. Now that we have a more or less clear idea of what things are being moved from A to B we can switch to explicit returns and signatures!